app/src/ui/repository.tsx

```diff
@@ -241,12 +247,71 @@ export class RepositoryView extends React.Component<
     }
   }
 
+  private currentStashForBranch() {
+    if (!enableStashing()) {
+      return null
+    }
+
+    const { branchesState, stashEntries } = this.props.state
+    const tip = branchesState.tip
+    if (tip.kind !== TipState.Valid) {
+      return null
+    }
+
+    // Using the short form ref name because the branch model does not keep track
+    // of the canonical ref name. Todo: update model to use the canonical ref name.
+    const branch = tip.branch
+    const stashEntry = stashEntries.get(branch.name)
+    if (stashEntry === undefined) {
+      return null
+    }
+
+    return stashEntry
+  }
+
   private renderContent(): JSX.Element | null {
     const selectedSection = this.props.state.selectedSection
-
     if (selectedSection === RepositorySectionTab.Changes) {
       const { changesState } = this.props.state
-      const { workingDirectory, selectedFileIDs, diff } = changesState
+      const {
+        workingDirectory,
+        selectedFileIDs,
+        diff,
+        shouldShowStashedChanges,
+      } = changesState
+
+      if (shouldShowStashedChanges && selectedFileIDs.length === 0) {
+        const stashEntry = this.currentStashForBranch()
+        if (stashEntry === null) {
+          return null
+        }
+
+        if (stashEntry.files.kind === StashedChangesLoadStates.Loaded) {
+          return (
+            <StashDiffViewer
+              stashEntry={stashEntry}
+              selectedStashedFile={
+                this.props.state.changesState.selectedStashedFile
+              }
+              stashedFileDiff={
+                this.props.state.changesState.selectedStashedFileDiff
+              }
+              imageDiffType={this.props.imageDiffType}
+              fileListWidth={this.props.stashedFilesWidth}
+              externalEditorLabel={this.props.externalEditorLabel}
+              onOpenInExternalEditor={this.props.onOpenInExternalEditor}
+              repository={this.props.repository}
+              dispatcher={this.props.dispatcher}
+            />
+          )
+        } else if (this.props.state.branchesState.tip.kind === TipState.Valid) {
+          this.props.dispatcher.loadStashedFiles(
```

Comment:
This is a dangerous pattern, we should not be calling the dispatcher from render methods. The Repository component could get re-rendered any number of times before the `loadStashedFiles` method is is able to complete and emit an update, causing us to shell out to Git for each render.

I think we can find a much better pattern for this by letting the appStore take care of loading stashed files when the right conditions exists for doing so but in the meantime I'm going to open a PR to ensure we're never loading stashed files more than once for any stash entry.

Reply:
I've opened https://github.com/desktop/desktop/pull/7267 as a stop-gap measure for this but I think we should address this more holistically before we ship to production.

Reply:
Merged.

Reply:
https://github.com/desktop/desktop/issues/7270 created to track this.